Add documentation for MCP Registry feature (#24713)
mlflow/mlflow: docs/docs/genai/mcp-registry/register-mcp-servers.mdx

---
title: Register MCP Servers
description: Learn how to register MCP servers in the MLflow MCP Registry using the UI and Python APIs, including server_json structure and tool auto-discovery.
---

import { APILink } from "@site/src/components/APILink";

# Register MCP Servers

This guide walks you through registering MCP servers in the MLflow MCP Registry. You can register servers from a `server_json` payload, from a URL, or through the UI.

## Register from a server_json Payload

> The most direct way to register an MCP server is by providing a `server_json` dictionary containing the server's configuration. Use <APILink fn="mlflow.genai.register_mcp_server" />:

Reviewer: Could you link to the schema? https://registry.modelcontextprotocol.io/docs#/schemas/ServerJSON

Ask your coding agent to find all other places that should reference the schema/open standard.

```python
import mlflow

version = mlflow.genai.register_mcp_server(
    server_json={
        "name": "io.github.anthropic/brave-search",
        "version": "1.0.0",
        "description": "Brave Search MCP server for web and local search",
        "icons": {
            "light": "https://example.com/icon-light.svg",
            "dark": "https://example.com/icon-dark.svg",

>         },

Reviewer: **🤖 AI Review**

This `icons` shape is rejected by registration. `_validate_mcp_icon_payloads` requires a **list** of objects with a `src` key (optional `theme`), not a `{light, dark}` map. Copy-pasting this example fails with `Invalid server_json.icons. Expected a list.`

```suggestion
        "icons": [
            {"src": "https://example.com/icon-light.svg", "theme": "light"},
            {"src": "https://example.com/icon-dark.svg", "theme": "dark"},
        ],
```

                "registryType": "npm",
                "identifier": "@anthropic/brave-search-mcp",
                "transport": {"type": "stdio"},
                "version": "1.0.0",
            }
        ],
        "remotes": [
            {
                "url": "https://mcp.example.com/brave-search",
                "type": "streamable-http",
            }
        ],
    },
    status="active",
    # Optional: record the source of this server definition
    source="https://github.com/anthropic/brave-search-mcp",
)

print(f"Registered '{version.name}' version {version.version}")
```

If the parent `MCPServer` does not yet exist, it is created automatically. If you register a new version under an existing server name, a new `MCPServerVersion` is added.

### server_json Structure

The `server_json` payload must contain `name` and `version` at the top level. Other fields are optional:

| Field | Required | Description |
|-------|----------|-------------|
| `name` | Yes | Namespaced server name (e.g., `io.github.anthropic/brave-search`) |
| `version` | Yes | Semantic version string (e.g., `1.0.0`, `2.0.0-beta.1`) |
| `description` | No | Human-readable description of the server |

> | `icons` | No | Object with `light` and/or `dark` icon URLs |

Reviewer: **🤖 AI Review**

Same issue as the example above: `icons` is a list of `{src, theme?}` objects, not a light/dark URL map.

```suggestion
| `icons` | No | List of icon objects with `src` and optional `theme` (`light`, `dark`, or omitted for any) |
```

```python
import mlflow

# From an HTTP URL
version = mlflow.genai.register_mcp_server_from_url(
    url="https://raw.githubusercontent.com/example/mcp-server/main/server.json",
    status="active",
)

# From a local file path
version = mlflow.genai.register_mcp_server_from_url(
    url="/path/to/server.json",
    status="draft",
)
```

The `source` field defaults to the URL when not explicitly provided, recording the provenance of the server definition.

## Register via the UI

![Register MCP Server UI](/images/llms/mcp-registry/register-mcp-server-ui.png)

1. Navigate to the **MCP Registry** section in the MLflow sidebar.
2. Click the **Create MCP server** button.
3. Enter a **Display name** for the server.
4. Edit the **server.json** payload in the editor with the server's `name`, `version`, `description`, `remotes`, and any other fields.
5. Select the initial **Status** and optionally provide a **Source** URL, icons, and tags.
6. Click **Create** to register.

## Tool Auto-Discovery

When you register a server without explicitly providing tools, MLflow automatically attempts to discover tools by connecting to the first usable remote URL in `server_json.remotes[]`.

:::note

> Tool auto-discovery requires the `mcp` extra: `pip install 'mlflow[mcp]'`. Discovery failure (network issues, auth requirements, timeouts) does **not** abort registration — the version is created with `tools=None`.

Reviewer: ```suggestion
Tool auto-discovery requires the `mcp` extra: `pip install 'mlflow[mcp]'`. Discovery failure (network issues, auth requirements, timeouts) does **not** abort registration. The version is still created with `tools=None`.
```

:::

You can control this behavior:

```python
import mlflow

# Let MLflow auto-discover tools (default behavior)
version = mlflow.genai.register_mcp_server(
    server_json={...},
)

# Skip tool discovery by passing an empty list
version = mlflow.genai.register_mcp_server(
    server_json={...},
    tools=[],
)

# Provide auth headers for tool discovery from a protected server
version = mlflow.genai.register_mcp_server(
    server_json={...},
    mcp_server_access_headers={"Authorization": "Bearer <token>"},

> )

Reviewer: **🤖 AI Review**

`server_json={...}` is not a dict — in Python that is a set containing `Ellipsis`. Please use a real minimal payload (or a named `server_json = {...}` variable) so the snippet is copy-pasteable.

Suggested rewrite:

```python
import mlflow

server_json = {
    "name": "io.github.anthropic/brave-search",
    "version": "1.0.0",
    "remotes": [
        {"url": "https://mcp.example.com/brave-search", "type": "streamable-http"},
    ],
}

# Let MLflow auto-discover tools (default behavior)
version = mlflow.genai.register_mcp_server(server_json=server_json)

# Skip tool discovery by passing an empty list
version = mlflow.genai.register_mcp_server(
    server_json=server_json,
    tools=[],
)

# Provide auth headers for tool discovery from a protected server
version = mlflow.genai.register_mcp_server(
    server_json=server_json,
    mcp_server_access_headers={"Authorization": "Bearer <token>"},
)
```
